scipy/optimize/_root_scalar.py

```diff
@@ -275,7 +275,21 @@ def root_scalar(f, args=(), method=None, bracket=None,
             raise ValueError('Bracket needed for %s' % method)
 
         a, b = bracket[:2]
-        r, sol = methodc(f, a, b, args=args, **kwargs)
+        try:
+            r, sol = methodc(f, a, b, args=args, **kwargs)
+        except ValueError as e:
+            # gh-17622 fixed some bugs in low-level solvers by raising an error
+            # (rather than returning incorrect results) when the callable
+            # returns a NaN. It did so by wrapping the callable rather than
+            # modifying compiled code, so the iteration count is not available.
+            if hasattr(e, "_x"):
+                sol = optzeros.RootResults(root=e._x,
+                                           iterations=np.nan,
+                                           function_calls=e._function_calls,
+                                           flag=str(e))
+            else:
+                raise
```

Comment:
Should use `raise from` and specify the error `e` no?

Reply:
If you want this line to appear in the traceback, yes. But I don't think this line should appear in the traceback. (That was my understanding anyway - I should test that.)

Reply:
```python3
def myfun():
    raise Exception

def myfun2():
    try:
        myfun()
    except Exception as e:
        raise
        # raise Exception from e

myfun2()
```
As is, we get:

<details>

```
Traceback (most recent call last):
  File "C:\Users\matth\anaconda3\envs\scipydev\lib\site-packages\IPython\core\interactiveshell.py", line 3442, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-dad59d4e01a3>", line 1, in <module>
    runfile('C:\\Users\\matth\\AppData\\Roaming\\JetBrains\\PyCharm2022.3\\scratches\\scratch_11.py', wdir='C:\\Users\\matth\\AppData\\Roaming\\JetBrains\\PyCharm2022.3\\scratches')
  File "C:\Program Files\JetBrains\PyCharm 2022.3.1\plugins\python\helpers\pydev\_pydev_bundle\pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2022.3.1\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:\Users\matth\AppData\Roaming\JetBrains\PyCharm2022.3\scratches\scratch_11.py", line 11, in <module>
    myfun2()
  File "C:\Users\matth\AppData\Roaming\JetBrains\PyCharm2022.3\scratches\scratch_11.py", line 6, in myfun2
    myfun()
  File "C:\Users\matth\AppData\Roaming\JetBrains\PyCharm2022.3\scratches\scratch_11.py", line 2, in myfun
    raise Exception
Exception
```

</details>

If you replace with the commented line, we get:

<details>

```
Traceback (most recent call last):
  File "C:\Users\matth\AppData\Roaming\JetBrains\PyCharm2022.3\scratches\scratch_11.py", line 6, in myfun2
    myfun()
  File "C:\Users\matth\AppData\Roaming\JetBrains\PyCharm2022.3\scratches\scratch_11.py", line 2, in myfun
    raise Exception
Exception
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "C:\Users\matth\anaconda3\envs\scipydev\lib\site-packages\IPython\core\interactiveshell.py", line 3442, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-dad59d4e01a3>", line 1, in <module>
    runfile('C:\\Users\\matth\\AppData\\Roaming\\JetBrains\\PyCharm2022.3\\scratches\\scratch_11.py', wdir='C:\\Users\\matth\\AppData\\Roaming\\JetBrains\\PyCharm2022.3\\scratches')
  File "C:\Program Files\JetBrains\PyCharm 2022.3.1\plugins\python\helpers\pydev\_pydev_bundle\pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2022.3.1\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:\Users\matth\AppData\Roaming\JetBrains\PyCharm2022.3\scratches\scratch_11.py", line 11, in <module>
    myfun2()
  File "C:\Users\matth\AppData\Roaming\JetBrains\PyCharm2022.3\scratches\scratch_11.py", line 9, in myfun2
    raise Exception from e
Exception
```

</details>

Which is more verbose but not very meaningful here. The intent is for the PR to modify the behavior only in this very particular situation that a NaN is encountered; the rest is supposed to be unchanged.

Reply:
I am with shipping like that. Just wanted to ensure it was not an oversight. Only reason to use the `raise from e` here would be to preserve the type of exception so that we can catch it a bit more selectively down the road.